### ISSUE TITLE
vmstorage merge rows operations stop queries from getting an answer

### ISSUE THREAD

**Issue by Irrational-NX** [VictoriaMetrics/VictoriaMetrics#2211]
**Describe the bug**
From time to time, Victoria Metrics tries to compact blocks of data (for performance reason I imagine). Here's the corresponding line of log :
`merged 457287132 rows across 850771 blocks in 67.389 seconds at 6785751 rows/sec to \"/vmstorage/data/data/big/2022_02/457190970_105474_20220215091812.477_20220217075903.124_16D219AA6395B943\"; sizeBytes: 281032367"`
The indicated number of rows/blocks and duration can vary a lot.

The problem, is that during this time, Grafana (which is configured with alertmanager and so does very frequent queries) can't get a response from VM.
This can be seen in VMSelect as well :
`slow query according to -search.logSlowQueryDuration=5s: remoteAddr="X.X.X.X:XXXXX", X-Forwarded-For: "X.X.X.X", duration=31.132 seconds; requestURI: "/select/573/prometheus/api/v1/query_range?blablabla&step=1"`

**To Reproduce**
Just wait, it happens from time to time, a few times per week.

**Expected behavior**
I do not expect this to happen, VM should always be available/joinable.
Since the merge operations are not very long, If I could schedule them to only happen during the night or the weekend, I could adapt my alerts in consequence.

**Version**
Version 1.71 for all components (vmstorage, vmwauth, vminsert, vmselect...)

**Comment by hagen1778** [comment]
Hi @Irrational-NX! 
The merging process is inevitable for [storage](https://docs.victoriametrics.com/Single-server-VictoriaMetrics.html#storage) health and performance:
* it improves query performance, since lower number of parts are inspected with each query;
* it reduces the number of data files, since each part contains fixed number of files;
* better compression rate for the resulting part.

In general, merging process left unnoticeable and may result into degradation only on systems with limited CPU resources or "slow" disks. However, you can control the number of allocated CPUs for merge operation via the following flag:
```
 -bigMergeConcurrency int
    	The maximum number of CPU cores to use for big merges. Default value is used if set to 0
```

Try it and let me know if it helps. I'd also recommend configuring [monitoring and alerting](https://docs.victoriametrics.com/Single-server-VictoriaMetrics.html#monitoring) in order to better understand what happens to your systems and resource usage.

**Comment by bs-amitmahajan29** [comment]
I'm getting the similar issue. Due to merging operation, the queries fired from vmselect aren't able to get any data due to timeouts.
(**PS**: _The timeout isn't due to heavy queries as these queries run very fast when merging operation isn't going on_)

Attaching the logs below for the 3 vmstorage pods with gp2 storageClass wherein i'm able to see that merging operation is going on and is taking too much time.
```
**pod1**
2023-09-02T04:28:23.211Z	info	VictoriaMetrics/lib/storage/partition.go:1366	merged (15 parts, 7168861 rows, 5985478 blocks, 105114752 bytes) into (1 part, 7103130 rows, 1173773 blocks, 27880239 bytes) in 2477.716 seconds at 2893 rows/sec to "/home/data/big/2023_09/7103130_1173773_20230901100045.232_20230902034135.000_1780DC53D680CC6A"
**pod2**
2023-09-02T04:28:58.090Z	info	VictoriaMetrics/lib/storage/partition.go:1366	merged (15 parts, 6967435 rows, 6036584 blocks, 100774272 bytes) into (1 part, 6430079 rows, 1190257 blocks, 31758600 bytes) in 1661.524 seconds at 4193 rows/sec to "/home/data/big/2023_09/6430079_1190257_20230901102155.475_20230902033106.559_1780DC5D5683B515"
**pod3**
2023-09-02T04:21:20.192Z	info	VictoriaMetrics/lib/storage/partition.go:1366	merged (15 parts, 7093793 rows, 6004990 blocks, 104851968 bytes) into (1 part, 6608982 rows, 1170276 blocks, 26337888 bytes) in 2424.143 seconds at 2926 rows/sec to "/home/data/big/2023_09/6608982_1170276_20230901100006.759_20230902034020.000_1780DC66E8EEF1FB"
```

I also increased the CPU and checked from the resourceUtilization for vmstorage pods that the CPU is available, yet the queries aren't getting through successfully.
<img width="887" alt="Screenshot 2023-09-02 at 10 44 20 AM" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/87753396/e0395766-5d2b-477b-8f74-3eafad5a66a6">

What can be the issue here and how can we fix the same?
**Note**: _Using vmstorage version installed with vmcluster tag_ `v1.85.3-cluster`

**Comment by hagen1778** [comment]
@bs-amitmahajan29 @Irrational-NX  can you show screenshots of Grafana dashboard for `ResourceUsage`, `Troubleshooting` and `vmstorage` panels?

**Comment by bs-amitmahajan29** [comment]
@hagen1778 attaching the details below.

### Details around the issue
Note: Using vmstorage version installed with vmcluster tag v1.85.3-cluster
T = Time at which merging operation started.
Sharing the graphs from T-2 hours, till T+2 day (merging operation hasn't ended as per timelines yet).

As per the below images, I'm able to see that the merging operation started at the month's start(i.e 0 UTC) and that was the time when CPU spiked for vmstorage and it was also the time when we could see in overview stats that queries started to fail due to timeout. (The drops in query duration in between are due to the pod restarts that we were doing either manually or to increase the CPU quota of the vmstorage pods)

**Resource Usage**

- After Issue Started
<img width="1408" alt="Screenshot 2023-09-04 at 1 49 19 PM" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/87753396/3fa42971-a2bb-4529-80ee-f1d63fed8cc5">
<img width="707" alt="Screenshot 2023-09-04 at 1 49 57 PM" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/87753396/5f0ad5ac-5a3e-4b63-93fc-f047581aa57e">

- Before the Issue
<img width="1409" alt="Screenshot 2023-09-04 at 2 09 22 PM" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/87753396/75df32a0-574b-48d6-9435-fd26b598efac">
<img width="735" alt="Screenshot 2023-09-04 at 2 09 37 PM" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/87753396/8184866e-672b-48e1-903e-95e586684ee9">


**Troubleshooting**
- After Issue Started
<img width="1406" alt="Screenshot 2023-09-04 at 1 51 08 PM" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/87753396/fd559b87-2586-40e4-afde-7ba42ecb364f">
<img width="1409" alt="Screenshot 2023-09-04 at 1 51 30 PM" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/87753396/9937cadd-cab4-4a0b-8978-6f4c425f798d">

- Before the Issue
<img width="1403" alt="Screenshot 2023-09-04 at 2 10 23 PM" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/87753396/338b15d9-4c1d-4d16-94cd-d9880881fef2">
<img width="1413" alt="Screenshot 2023-09-04 at 2 10 40 PM" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/87753396/682e8ae4-28a5-4494-8cbd-5f519188fccc">



**vmstorage**
- After Issue Started
<img width="1408" alt="Screenshot 2023-09-04 at 1 52 14 PM" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/87753396/da887afe-b19a-47f6-9f90-e08524d41ae4">
<img width="1411" alt="Screenshot 2023-09-04 at 1 52 30 PM" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/87753396/0dc265fb-d72d-4171-8ece-8eff8f537d9a">

- Before the Issue
<img width="1409" alt="Screenshot 2023-09-04 at 2 11 25 PM" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/87753396/09a8bc26-9ddb-42ef-8b75-40941c479c68">
<img width="1411" alt="Screenshot 2023-09-04 at 2 11 42 PM" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/87753396/7ad4abb2-6c56-4b3f-b5db-ce17568491cf">



**Overview stats**
Also sharing the stats of query duration (_FYI: We have kept the query timeout as 10 sec_)
- After Issue Started
<img width="1408" alt="Screenshot 2023-09-04 at 1 55 40 PM" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/87753396/6dc3a87e-adcb-4501-9a2c-c19cba0fb1d8">

- Before the Issue
<img width="1412" alt="Screenshot 2023-09-04 at 2 12 17 PM" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/87753396/08426013-e2e8-46d9-b165-a5ac007c3623">



### Steps that we tried doing
1) We tried restarting manually the pod(s) of vmstorage to check if it was due to any process getting stuck or something, but even after doing so we could observe that merging operation was still resuming which we concluded as it wasn't any process stuck issue as such.
2) We then tried increasing the CPU requests and limit of vmstorage pods gradually till the thresholds of our nodeGroup(_r6i.large_) reached.
_Note_: As seen from the "_Before the Issue_" section in above screenshots, vmstorage usually can be seen consuming very less CPU.
3) We also changed the PVC storage type from gp3 to gp2 but that didn't help either.
3) We then shifted the vmstorage pods to r6i.xlarge nodeGroups so that we'd be able to add even more CPU to the pods, and post doing that the querying was possible during the merging operation.


### Additional doubts
1) Is there any way through which I can limit the CPU occupied by merging operation so that querying isn't impacted? Since merging is a background operation I'm fine with it taking a longer amount of time but I'm not fine with it impacting the querying over the cluster.

**Comment by hagen1778** [comment]
@bs-amitmahajan29 thanks for provided screenshots!
A few questions:
1. Do you use [forced merge](https://docs.victoriametrics.com/#forced-merge)?
2. Do you use [downampling](https://docs.victoriametrics.com/#downsampling)?
3. What kind of service generates read requests in this setup?

> Is there any way through which I can limit the CPU occupied by merging operation so that querying isn't impacted? 

Yes, you can limit amount of parallel workers doing the background merge by adjusting `bigMergeConcurrency` cmd-line flag on vmstorage level to the lower values. Please note, this flag was deprecated in newer versions.

From the provided screenshots I see that when the issue started, the amount of goroutines and read request rate jumped significantly. For me this is a sign that client issuing read requests generates a thundering herd problem. Once the vmstorage started to fall behind with requests serving, the requests queue starts to build up. The queue procession then consumes all the resource left in the cluster, because it now needs to serve much more concurrent read requests than before. I believe, the read requests queue increased the duration of incident significantly.

I'm also interested in getting responses to questions above.

**Comment by bs-amitmahajan29** [comment]
**Regarding the mentioned questions above, answering them below.**
1) The merging happened as usual on the month start date as seen from the screenshots. However, when we tried out the things like killing of pod and scaling up resources we initiated the force merge process by hitting `internal/force_merge` endpoint on vmstorage.
2) We are not using downsampling as it's an enterprise grade feature.
3) Read requests are generated majorly from grafana, vmalert and few requests are generated from custom scripts which query via vmauth (the frequency for such scripts is quite low as compared to the frequency via grafana and vmalert).
_Note: The requests via grafana and vmalert both come via vmauth itself._
Also, when the incident started on month start, no new additional requests as such were generated.


**Regarding the `bigMergeConcurrency` cmd-line flag,**
I could see that the values that this flag supports is integer type, so does it mean that the lowest value that I can give it is 1? Also, I had tried setting it as `1`, but I still could see that by using it the CPU requirement during merging process was the same. Also, what is the default value that flag if not provided (I could [see that default is used](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/v1.85.3-cluster/app/vmstorage/main.go#L41) if not explicitly provided, but what exactly would be default in this case)?
_Note: I'm using 3 vmstorage pods with 2 replication factor._ and below attached resources as of now.
<img width="185" alt="Screenshot 2023-09-05 at 8 58 34 PM" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/87753396/9362d106-837e-49a7-9248-a0d46840eb8d">

**Few additional questions**
1) I don't think it's due to read requests increase as such because no additional read requests were generated from the above mentioned sources. Can it be the merging process itself that generated read requests on the disk?
2) What would you suggest as the best possible tuning through `bigMergeConcurrency` and `smallMergeConcurrency` that can be done to make sure that I don't have to keep the CPU to such high values for vmstorage? I'm fine with the merging operation taking long as far as it doesn't impact querying.
3) Any way through which I can disable merging feature itself? The retention of data in the cluster is 90 days and I'm okay if the merging doesn't happen at all (as far as the data gets deleted after retentionPeriod, even if merging doesn't happen I'm okay)
4) I'm currently on `v1.88.3-cluster` version. What version would you recommend me to go on? (Assuming that upgrading to newer version would help in resolving the issue)

**Comment by hagen1778** [comment]
>  Can it be the merging process itself that generated read requests on the disk?

No, merge process doesn't generate any HTTP requests. vmalert and grafana do no retries as well, afaik. vmauth can do retries but only to network errors, not when server responded with slow query or other error.

> What would you suggest as the best possible tuning through bigMergeConcurrency and smallMergeConcurrency that can be done to make sure that I don't have to keep the CPU to such high values for vmstorage?

I'd recommend tuning only `bigMergeConcurrency` flag and try reducing it by x2 as start. This flag controls the amount of CPU threads doing the merge in parallel. The more threads there are, the faster merge is going, but more CPU and IO resources will be consumed. By default, bigMergeConcurrency is limited by the number of available CPU cores available for the process. On instances with slow disks merge process can saturate disk throughput and lead to excessive CPU usage and slow queries. 

> Any way through which I can disable merging feature itself?

No, merging can't be disabled. The monthly merge like in your case usually happens if deduplication or downsampling is enabled, or if force merge was called. Its performance can be limited by reducing  `bigMergeConcurrency`  flag value.

> I'm currently on v1.85.3-cluster version. What version would you recommend me to go on? (Assuming that upgrading to newer version would help in resolving the issue)

I'd recommend to update to LTS version: https://docs.victoriametrics.com/CHANGELOG.html#v1878 or https://docs.victoriametrics.com/CHANGELOG.html#v1933. 
Please note, in https://docs.victoriametrics.com/CHANGELOG.html#v1933 `bigMergeConcurrency` flag was deprecated:
> FEATURE: deprecate -bigMergeConcurrency command-line flag, since improper configuration for this flag frequently led to uncontrolled growth of unmerged parts, which, in turn, could lead to queries slowdown and increased CPU usage. The concurrency for [background merges](https://docs.victoriametrics.com/#storage) can be controlled via -smallMergeConcurrency command-line flag, though it isn't recommended to change this flag in general case.

Also, the https://docs.victoriametrics.com/CHANGELOG.html#v1933 has breaking changes on storage layout. So rolling back from 1.90+ to 1.89- is impossible.

-----------------------

I'd recommend updating to https://docs.victoriametrics.com/CHANGELOG.html#v1878 and observe if issue still remains. If you will be able to reproduce the issue once again and observe that number of goroutines increased for vmstorage in the same way - please [collect and send goroutines profile](https://docs.victoriametrics.com/Cluster-VictoriaMetrics.html#profiling) (the path for goroutines profile is `/debug/pprof/goroutine`) from vmstorage during the incident. The profile could help us to debug the problem

**Comment by bs-amitmahajan29** [comment]
@hagen1778 I'm still not sure as to how would switching over to new LTS solve this issue? I checked and found that the logic for mergeWorkers and CPU allotment hasn't changed as such between the versions.

I'm fine with merging happening for performance purposes, but I just want to make sure that it doesn't consume too much CPU.
On usual days (when merge operation isn't going on), the vmstorage consumes on an average of 0.2 CPU per pod as attached in screenshot below.
<img width="1420" alt="Screenshot 2023-09-09 at 6 41 42 PM" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/87753396/9a026a40-5272-4897-bad3-4c34900eb0ab">
<img width="696" alt="Screenshot 2023-09-09 at 6 41 23 PM" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/87753396/85433c80-69b6-495e-8874-fed205fbb0ed">

I also checked that when merge operation is going on, the CPU spikes upto a maximum of ~1.2 per pod as seen from below screenshots. Still for some reason, I have to give ~2.2 CPU per pod for merging and querying to both work as expected (and yet the actual usage doesn't go beyond the ~1.2 CPU). So i'm still wondering that why the extra unused CPU needs to be allocated per vmstorage for it to function for both querying and merging?
<img width="1376" alt="Screenshot 2023-09-09 at 6 44 43 PM" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/87753396/78c319e8-1122-4e4e-a200-5c26fc86d53e">
<img width="704" alt="Screenshot 2023-09-09 at 6 43 05 PM" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/87753396/55e6c40f-5e8f-4309-b2ea-fdbfa156cf82">

**Comment by hagen1778** [comment]
> @hagen1778 I'm still not sure as to how would switching over to new LTS solve this issue? I checked and found that the logic for mergeWorkers and CPU allotment hasn't changed as such between the versions.

Updating to LTS means running the latest support version. Version where we can deliver code changes if something is wrong with the code. Although, there is always a chance the newer version has addressed or changed a logic already.

> I'm fine with merging happening for performance purposes, but I just want to make sure that it doesn't consume too much CPU.
On usual days (when merge operation isn't going on), the vmstorage consumes on an average of 0.2 CPU per pod as attached in screenshot below.

The problem you mentioned isn't the lack of CPU, it is the disk throughput limit. From the screenshots it looks like disks are limited with 100MB/s read/write speed. Once disk bandwidth is saturated, CPU starts getting overused to compensate slow disks - the caches can't update entries, the page cache gets polluted, etc. I believe that once disks aren't a bottleneck - no CPU issues will appear. Do you run a nodeexporter on vmstorage instances? Can we check the following query during incident? 
```
sum by(job, instance) (irate(node_cpu_seconds_total{mode="iowait"}[5m])) / on(instance) group_left sum by (instance)((irate(node_cpu_seconds_total{}[5m])))
```


>  Can it be the issue that vmselect which issues TCP connection to vmstorage keeps on trying and some kind of lock mechanism gets kicked in? And these many TCP connections result into problem? (Attaching screenshot below)

There shouldn't be such issue. At least, I didn't see it before and code implementation suggests it should be there. That's why I ask to collect profile during incident:
> If you will be able to reproduce the issue once again and observe that number of goroutines increased for vmstorage in the same way - please [collect and send goroutines profile](https://docs.victoriametrics.com/Cluster-VictoriaMetrics.html#profiling) (the path for goroutines profile is /debug/pprof/goroutine) from vmstorage during the incident. The profile could help us to debug the problem

**Comment by bs-amitmahajan29** [comment]
```
Do you run a nodeexporter on vmstorage instances? Can we check the following query during incident?
```
@hagen1778 We have nodeexporter running in our cluster. Sharing the screenshot for the same below before and during the incident.

![Screenshot 2023-09-14 at 1 00 53 PM](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/87753396/6da4da66-7a4d-45c5-bad0-f28c46fa003e)

**Comment by hagen1778** [comment]
Hi @bs-amitmahajan29! Sorry for late reply.
I think the screenshot shows IO issues during the incident - CPUs were waiting for disks to respond with data. 
Have you tried following recommendations above?

**Comment by bs-amitmahajan29** [comment]
Hi @hagen1778, no worries 💯 
```
Have you tried following recommendations above?
```
I noticed this issue today as well in vmstorage component. Attaching the profile of vmstorage. ([Drive Link](https://drive.google.com/drive/folders/151WXau3wDHOcINB8r5b8GcfFfAE0lWI0))

Regarding the upgrade to LTS version, we are evaluating this as well, but in parallel can you help us with the profile results?

**Comment by rohit3463** [comment]
Hello @hagen1778 

In the graph 

> ```
> Do you run a nodeexporter on vmstorage instances? Can we check the following query during incident?
> ```
> 
> @hagen1778 We have nodeexporter running in our cluster. Sharing the screenshot for the same below before and during the incident.
> 
> ![Screenshot 2023-09-14 at 1 00 53 PM](https://user-images.githubusercontent.com/87753396/267894423-6da4da66-7a4d-45c5-bad0-f28c46fa003e.png)

It looks like CPU is waiting but still I think it should be able to serve queries, but the problem here is the TCP connections. 


> @hagen1778 one more thing which I have noticed is that,
> 
>     * When merging starts in the cluster, at a point of time the total number of TCP connections spike alot. Can it be the issue that vmselect which issues TCP connection to vmstorage keeps on trying and some kind of lock mechanism gets kicked in? And these many TCP connections result into problem? (Attaching screenshot below)
> 
> 
> (Possibly related to : [#2552 (comment)](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/2552#issuecomment-1383382741))
> 
> (As seen from image, the timeframe during which the goroutines can be seen increasing matches with the timeframe at which we could see a huge spike in TCP connections. However, even when this is happening, the CPU can be seen hovering at around 45-50% but still it results into connections not being made vmselect to vmstorage as the query fail during the same time)
> <img alt="Screenshot 2023-09-11 at 12 25 58 PM" width="1368" src="https://user-images.githubusercontent.com/87753396/266933143-d82547b1-f52a-4bdc-84a9-1919ccf74c4d.png"> <img alt="Screenshot 2023-09-11 at 12 31 04 PM" width="708" src="https://user-images.githubusercontent.com/87753396/266934317-6222cde1-6c47-41eb-ac7a-c35b0caa0665.png">

As we can see in the screenshot, the number of TCP connections have increased drastically which should not be related to CPU wait of disk IO restrictions. This is the major problem that is blocking all the queries. Is this expected?

**Comment by ruslanboyarsky** [comment]
> Hi, Looks like we observe similar issue in our VM cluster (v1.87.6-cluster).
> 
> Spontaneously, random vmstorage node of our VM cluster start to have the following symptoms:
> 
> 1. LA is grown:  ` value > 100`
> 2. we see a lot of open connects. `ss -nar | wc -l  > 40000`
> 3. amount of gorouting is grown:  `> 40000`
> 4. Open TCP connections of vmstorage is grown as well
>    The node starts to slow process any incoming requests and in results, inserts start to process incoming requests slowly. As result, throughput of out VM cluster is degraded
> 
> ![image](https://user-images.githubusercontent.com/112613645/269683075-c3095d6b-14a2-4351-b7bd-c9fda9f49705.png) ![image](https://user-images.githubusercontent.com/112613645/269683191-e48bb4e3-b809-4862-8734-c0652fc28858.png)
> 
> As well, I've collected cpu and goroutins profiles: [pprof.zip](https://github.com/VictoriaMetrics/VictoriaMetrics/files/12688539/pprof.zip)
> 
> Let me know if you need more info from my side

I've crated separated issue https://github.com/VictoriaMetrics/VictoriaMetrics/issues/5045

**Comment by hagen1778** [comment]
Hello @rohit3463! Thanks for profile!
It shows that vmselect opened many connections to vmstorage, the `search.maxConcurrentRequests` limit was reached on vmstorage and all connections (goroutines) were put on hold.
What is the `search.maxConcurrentRequests` limit on vmselect in your case?

**Comment by hagen1778** [comment]
And how many vmselects do you have?

**Comment by bs-amitmahajan29** [comment]
2 VMSelects

**Comment by hagen1778** [comment]
@bs-amitmahajan29 can I request screenshots for the following panel during the issue:
1. Resource usage: RSS, RSS anon, CPU, Disk write/read, Disk calls
2. Troubleshooting section all
3. Overview: requests rate, query latency, ingestion, active time series
4. vmselect: concurrent selects, network usage
5. vminsert: storage connection saturation
?

I'll check why vmselect opens so many connections. If you say it is 20 concurrent reqs allowed per vmselect, then it shouldn't have more than 20*2 connections opened per vmstorage. 
I also don't think connections causing the delay. I think it is vmstorage starting to slow-downs because either insufficient resources or because of some expensive query which saturates resources. I hope requested screenshots will help me to understand the issue.
7.

**Comment by bs-amitmahajan29** [comment]
@hagen1778 I have already shared the above asked things in the following comments above, please let me know if I have missed any!
- https://github.com/VictoriaMetrics/VictoriaMetrics/issues/2211#issuecomment-1704904521
- https://github.com/VictoriaMetrics/VictoriaMetrics/issues/2211#issuecomment-1713292620

Regarding `because of some expensive query which saturates resources`, I don't think so that's the case because once we kill the pod(thereby terminating the merging operation) the queries run fine. Also, even the simplest queries like `/select/1/prometheus/api/v1/label/__name__/values` fail during the incident.

**Comment by hagen1778** [comment]
@bs-amitmahajan29 thanks! Wanted to get actual screenshots. If those are related to the latest observed incident - then ok.
Btw, they're missing `vmselect: concurrent selects, network usage`, `vminsert: storage connection saturation`.

Anyway, I see caches were reset multiple time here :
<img width="481" alt="image" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/2902918/a22c70a5-58e7-4141-981f-8b95ea0d01f2">
do you know why? was vmstorage killed ?

**Comment by hagen1778** [comment]
Thanks!
Sorry that I'm asking or repeating the same information again and again - just fell out of context since last time.
Let me sum up the observation:
1. Once per weak or frequently, vmcluster starts to lag: read queries become unresponsive, resource usage increases
2. The cluster config is the following: 3 vmstorage nodes with 2vCPUm 5GiB RAM, 2 vmselects
3. When this happens, vmstorage logs message about starting the merge
4. The merge duration can take many hours
5. Restart helps for a bit, until the merge starts again
6. Once merge is done, cluster becomes stable
Is this correct?

Checking profile from [this comment](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/2211#issuecomment-1727943567)  I see that significant amount of CPU time is spent on deduplication process. Deduplication process is similar to merge process - vmstorage takes a couple of data parts and creates a new single part, removing duplicates in the process. Deduplication process on storage is CPU and IO intensive: data blocks need to be read from disk (IO), unpacked and merged (CPU), and written back to disk (IO).

Deduplication process happens if `-dedup.minScrapeInterval` cmd-line flag was set. The deduplication process is limited with one thread to not affect the system significantly.

I think that because vmstorage has only 2vCPU allocated, deduplication takes much longer than expected. The deduplication slowness can also be caused by slow disks. So if one vCPU is busy with deduplication work and saturates disk, then the remaining resources capacity isn't enough for processing incoming read/write load: which makes read queries to pile up, and increases vminserts connection saturation.

I think the following steps need to be taken to reduce probability of such incidents:
1. CPU/mem requests and limits need to be set to equal values, to reduce frictions on resource shortage
2. The CPU limit for vmstorage need to be increased
3. Alternatively to p2, it is better to scale vmstorages from 3 to 5. This will increase amount of CPUs and disk IO bandwidth, making cluster more stable.
4. Disable deduplication of none of the above is acceptable

**Comment by valyala** [comment]
I suppose the following sequence of events could be the root cause of the significant slowdown in both data ingestion rate querying, plus the increased CPU usage at `vmstorage` nodes:

1. At the beginning of every month `vmstorage` starts final [deduplication](https://docs.victoriametrics.com/#deduplication) and [downsampling](https://docs.victoriametrics.com/#downsampling) for the data from the previous months. This process basically read all the data from the previous month, de-duplicate it and write the de-duplicated data back to disk. This process runs in a background thread, and can occupy up to 100% of a single CPU core.
2. `vmstorage` limits the number of background merges to the number of CPU cores (the de-duplication process is also counted as background merge). So, when the de-duplication process runs, other [background merges](https://docs.victoriametrics.com/#storage) can run on very limited number of CPU cores if `vmstorage` runs in an environment with CPU limits set to 1-2. This means that background merges start lagging behind the data ingestion rate.
3. This results in significant increase in the number of parts on disk (see [these docs](https://docs.victoriametrics.com/#storage) for information about parts), since they cannot be merged in a timely manner.
4. High number of parts lead to queries slowdown, since every query usually needs to inspect every newly created part on disk (if this part contains samples on the requested time range). This also increases CPU usage. The [official Grafana dashboard for VictoriaMetrics cluster](https://docs.victoriametrics.com/cluster-victoriametrics/#monitoring) contains a graph showing the number of parts in the database.
5. The increased CPU usage slows down background merges, including the de-duplication process for the previous month.
6. This leads to very long performance degradation.

There are the following workarounds:

- To increase CPU limits at `vmstorage` nodes, so background merges could keep up with the data ingestion rate when de-duplication process for the previous month is running or when some big merge is running. This should help reducing the number of parts in the storage, which, in turn, should increase query performance and reduce CPU usage.
- To increase the available disk read/write IO bandwidth at `vmstorage` nodes, so the de-duplication process or big background merge doesn't occupy all the available disk IO bandwidth.
- To remove `-smallMergeConcurrency` and `-bigMergeConcurrency` command-line flags from `vmstorage`, since they are deprecated. Too low values for these flags may lead the sequence of events described above.
- To remove other non-essential command-line flags from all the cluster components - vminsert, vmselect and vmstorage. There are non-zero chances that some flags have improper values, which can result in various issues similar to this one.
- To upgrade to [the latest release](https://docs.victoriametrics.com/CHANGELOG.html). It contains some improvements to background merge process, which should reduce the probability of the issue described above - see bb7a419cc30afed0a42c431d48cb5f4d1ac29e7b for details.
